### ISSUE TITLE
Interop clients should not override authority by default, for proper SNI testing

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc#16759]
You have to specify `GRPC_SSL_TARGET_NAME_OVERRIDE_ARG` to get SNI. That's dangerous as it may prevent server operators from presenting correct certificates, as is the case for `grpc-test.sandbox.gogole.com`.

Reproduction case:

```
ejona@bruteforce:~/clients/grpc-c$ docker_image=gcr.io/grpc-testing/grpc_interop_python:v1.15.0
ejona@bruteforce:~/clients/grpc-c$ docker run -i --rm=true -e PYTHONPATH=/var/local/git/grpc/src/python/gens -e LD_LIBRARY_PATH=/var/local/git/grpc/libs/opt -w /var/local/git/grpc --net=host $docker_image bash -c "py27_native/bin/python src/python/grpcio_tests/setup.py run_interop --client --args=\"--server_host=grpc-test.sandbox.googleapis.com --server_host_override=grpc-test.sandbox.googleapis.com --server_port=443 --use_tls=true --test_case=empty_unary\""
running run_interop
ejona@bruteforce:~/clients/grpc-c$ docker run -i --rm=true -e PYTHONPATH=/var/local/git/grpc/src/python/gens -e LD_LIBRARY_PATH=/var/local/git/grpc/libs/opt -w /var/local/git/grpc --net=host $docker_image bash -c "py27_native/bin/python src/python/grpcio_tests/setup.py run_interop --client --args=\"--server_host=grpc-test.sandbox.googleapis.com --server_port=443 --use_tls=true --test_case=empty_unary\""
running run_interop
Traceback (most recent call last):
  File "src/python/grpcio_tests/setup.py", line 107, in <module>
    test_runner=TEST_RUNNER,
  File "/var/local/git/grpc/py27_native/local/lib/python2.7/site-packages/setuptools/__init__.py", line 140, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/var/local/git/grpc/src/python/grpcio_tests/commands.py", line 188, in run
    self.run_client()
  File "/var/local/git/grpc/src/python/grpcio_tests/commands.py", line 204, in run_client
    client.test_interoperability()
  File "/var/local/git/grpc/src/python/grpcio_tests/tests/interop/client.py", line 127, in test_interoperability
    test_case.test_interoperability(stub, args)
  File "/var/local/git/grpc/src/python/grpcio_tests/tests/interop/methods.py", line 482, in test_interoperability
    _empty_unary(stub)
  File "/var/local/git/grpc/src/python/grpcio_tests/tests/interop/methods.py", line 145, in _empty_unary
    response = stub.EmptyCall(empty_pb2.Empty())
  File "/var/local/git/grpc/py27_native/local/lib/python2.7/site-packages/grpc/_channel.py", line 532, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/var/local/git/grpc/py27_native/local/lib/python2.7/site-packages/grpc/_channel.py", line 466, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Connect Failed"
	debug_error_string = "{"created":"@1538496521.874311362","description":"Failed to create subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":2636,"referenced_errors":[{"created":"@1538496521.874124955","description":"Pick Cancelled","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":241,"referenced_errors":[{"created":"@1538496521.874069460","description":"Connect Failed","file":"src/core/ext/filters/client_channel/subchannel.cc","file_line":663,"grpc_status":14,"referenced_errors":[{"created":"@1538496521.873898314","description":"Peer name foo.test.google.fr is not in peer certificate","file":"src/core/lib/security/security_connector/security_connector.cc","file_line":880}]}]}]}"
```

I did confirm with wireshark that SNI is used with the override and not used without.

**Comment by ejona86** [comment]
As part of this issue, I expect all the interop tests hitting `grpc-test.sandbox.gogole.com` to _not_ use `--server_host_override`. That should be verified cross-language.

**Comment by ejona86** [comment]
@jiangtaoli2016, the first time I ran it had the override and worked. The second time it didn't have the override and it failed. I showed that _I can't_ remove the override because otherwise it fails because SNI is missing. But also, I didn't make the first command line, it came from what is used for interop testing: https://github.com/grpc/grpc/blob/2d05737/tools/interop_matrix/testcases/python__master

**Comment by jiangtaoli2016** [comment]
It looks like server_host_override ("foo.test.google.fr") is used as default for interop test. This is the intended behavior for testing, since our interop server does not have "googleapi.com" certificate.

For c++ interop client, the default server_host_override is here
https://github.com/grpc/grpc/blob/master/test/cpp/interop/client.cc#L41

**Comment by mehrdada** [comment]
We pass channel args directly to core. The Python layer does not know anything about them. We don't even provide string constants.

**Comment by ejona86** [comment]
> It looks like server_host_override ("foo.test.google.fr") is used as default for interop test. This is the intended behavior for testing, since our interop server does not have "googleapi.com" certificate.

It does seem that foo.test.google.fr is the default override for C++, which is broken. The [spec](https://github.com/grpc/grpc/blob/v1.15.0/doc/interop-test-descriptions.md) is clear:

> --server_host_override=HOSTNAME
The server host to claim to be connecting to, for use in TLS and HTTP/2 :authority header. If unspecified, the value of --server_host will be used

And "the value of server_host will be used" is really just another way of saying, "we won't use any override."

Interop tests hitting a server locally should explicitly specify --server_host_override if it is necessary. The "default" for the interop client should be "normal" TLS with default handling of the authority (so implicit from the target/host) and using the system-default CA list.

> Same for python interop client
https://github.com/grpc/grpc/blob/v1.15.1/src/python/grpcio_tests/tests/interop/client.py#L55
Python interop test client will always add server_host_override during testing.

Hmmm... That's certainly weird, because I didn't see any SNI with wireshark when I didn't specify server_host_override. Maybe it was there and I just missed it. I can't easily re-test now.

In any case, the interop clients should only specify grpc.ssl_target_name_override when server_host_override is provided on the command line. Otherwise we aren't testing the client's behavior as a user would see.

**Comment by jiangtaoli2016** [comment]
+ @apolcyn @jtattermusch interop folks

@ejona86 Let's decouple of SNI from grpc interop tests. In grpc interops (c++, python, php at least). I have not checked Java/Go interop tests. Interop client will by default use foo.test.google.fr for hostname override and **always** override targetname. There is no way to not override. In other words, if you don't want override, you have to manually set ssl_target_name_overrides to be the same as the server_host. If it is something we don't like, let @apolcyn and @jtattermusch fix the interop tests.

In the case of non-interop-testing, users are not required to set grpc.ssl_target_name_override. It won't be set by default. If grpc.ssl_target_name_override is set, then SNI will be kicked in. Otherwise, there won't be SNI.

**Comment by jiangtaoli2016** [comment]
by the way, I agree 
https://github.com/grpc/grpc/blob/v1.15.0/doc/interop-test-descriptions.md should be updated with regarding to server_host_override.

**Comment by ejona86** [comment]
> It won't be set by default. If grpc.ssl_target_name_override is set, then SNI will be kicked in. Otherwise, there won't be SNI.

That is unacceptable. SNI is necessary for normal operation. It isn't really an "optional" feature in today's web. Users should normally only need to set the "target" string to configure their machine to point at a particular server; forcing them to propagate a second string to properly configure TLS is a usability issue.

To me, the interop tests are part of this discussion since they are supposed to be used to test that SNI is working.

**Comment by ejona86** [comment]
> by the way, I agree
https://github.com/grpc/grpc/blob/v1.15.0/doc/interop-test-descriptions.md should be updated with regarding to server_host_override.

I didn't mention anywhere that the document should be updated. What update do you think is appropriate?

**Comment by jiangtaoli2016** [comment]
In Java interop test TestServiceClient.java, if --server_host_override is not set, it won't use "foo.test.google.fr" as default. We'd better make interop consistent in all languages before update the interop-test-descriptions-md.

**Comment by ejona86** [comment]
> In Java interop test TestServiceClient.java, if --server_host_override is not set, it won't use "foo.test.google.fr" as default. 

Yes, which is what interop-test-descriptions.md says to do... In that case Java does not provide any override, which will compute the string to use for SNI based on the hostname/target string.

I had asked "What update do you think is appropriate?" and I still don't see you mention what change you thought may need to be made.

> We'd better make interop consistent in all languages before update the interop-test-descriptions-md.

No. That's the tail wagging the dog. If there are any issues the .md should be updated and the languages should be made to conform. As it stands, at least Python and C++ are not conforming.

**Comment by jiangtaoli2016** [comment]
PS: PHP have same the behavior as Python and C++ on interop test wrt server_host_override.

I am less concerned about interop test. I want to make sure we are align on SNI on non-testing environment. Let's chat about it.

**Comment by jiangtaoli2016** [comment]
> That is unacceptable. SNI is necessary for normal operation. It isn't really an "optional" feature in today's web. Users should normally only need to set the "target" string to configure their machine to point at a particular server; forcing them to propagate a second string to properly configure TLS is a usability issue.

I am not sure what usability you prefer. Could you clarify?

IMO, there are three separate issues
1. In grpc wrapped languages, hostname/authority is automatically plumbed for server name check and port is stripped. Most use case does **not** require SNI. If user wants to connects to 'prod.abc.com', but server only has a certificate on 'test.bcd.com', client can use SNI to configure 'test.bcd.com' using grpc.ssl_target_name_override. In my view, it is a very corner use case except for testing. Especially we have SPIFFE certificate coming out soon that user can have full control on server name check with a callback function (mainly for non-https semantics). If such way of passing SNI is not acceptable, what is the better way.

2. It appears a bug in python, that a user could not use SNI properly. We need to reproduce and see whether it is indeed a bug or user did not configure credential properly.
https://groups.google.com/d/msgid/grpc-io/b763a855-06b5-4c96-9a8d-2aca2b314802%40googlegroups.com

3. Interop-client. In c++, php, python, interop-client will always use SNI. And if there is no hostname overrides from command line, "foo.test.google.fr" will be used as hostname_overrides. It is different behavior from Java interop client and inconsistent with https://github.com/grpc/grpc/blob/v1.15.0/doc/interop-test-descriptions.md. I will defer interop folks (@jtattermusch @apolcyn ) to fix.

**Comment by ejona86** [comment]
> Most use case does not require SNI. If user wants to connects to 'prod.abc.com', but server only has a certificate on 'test.bcd.com', client can use SNI to configure 'test.bcd.com' using grpc.ssl_target_name_override.

SNI != testing. In fact, our usage is the only time I've seen SNI necessary for testing. The _client_ generally doesn't know if it needs to use SNI. Also, it may not need to use SNI today, but needs to use SNI tomorrow as the _server_ infrastructure changes. It is wrong to require the client to do additional configuration to enable SNI.

I'd agree that "most service _deployments_" do not require SNI, because SNI isn't needed with microservices. But that isn't the case for _hosted_ services; things like googleapis.com. It's very normal to need SNI in those environments. Any place virtual hosting takes place (which means an L7 LB, and maybe more rarely L4 LB) SNI should be used by clients.

> Especially we have SPIFFE certificate coming out soon that user can have full control on server name check with a callback function (mainly for non-https semantics).

"Here is a use case that doesn't need SNI" is not an argument against "we need SNI."

> It appears a bug in python, that a user could not use SNI properly.

This bug is not for the reported Python problem. I tested the interop client and saw SNI working. So something else is at play there.

> Interop-client. In c++, php, python, interop-client will always use SNI. And if there is no hostname overrides from command line, "foo.test.google.fr" will be used as hostname_overrides.

Which is just wrong and why I was not aware of this bug. The interop client is testing interopability, which includes using SNI. The fact that the interop tests are not testing _the default configuration_ is wrong and broken. It sounds like they always specify the override because SNI is not on by default; that is, they are working around the broken SNI in the interop client instead of fixing the gRPC library.

The interop client is important to this fix because it is the "test" that SNI is working correctly. This shouldn't be considered "fixed" without a test, and I expect that the interop test _should_ be that test. I'm fine if the one client is fixed (say, C++) and then other issues opened for the other clients. But I'm going to assume that implementations are broken until their interop client (which is something I can easily run) shows that it is working correctly.

**Comment by srini100** [comment]
Had a discussion with @jiangtaoli2016. Agreed that SNI should be enabled by default.

**Comment by jiangtaoli2016** [comment]
@ejona86 could you provide a detailed reproduce instruction? I just tried on my local workstation on my own ssl greeter example, without `GRPC_SSL_TARGET_NAME_OVERRIDE_ARG`, the SNI is set correctly as the target name of the server.

**Comment by ejona86** [comment]
I just tested the python interop client without the override arg and it does appear to have SNI. I may have been thrown off because it was defaulting to `foo.test.google.fr`.

So maybe only the interop clients need to be fixed to avoid setting GRPC_SSL_TARGET_NAME_OVERRIDE_ARG by default.

```
docker_image=gcr.io/grpc-testing/grpc_interop_python:v1.15.0
docker run -i --rm=true -e PYTHONPATH=/var/local/git/grpc/src/python/gens \
    -e LD_LIBRARY_PATH=/var/local/git/grpc/libs/opt -w /var/local/git/grpc --net=host \
    $docker_image bash -c \
    "sed -i 's/grpc.ssl_target_name_override/nothing/' src/python/grpcio_tests/tests/interop/client.py; \
    py27_native/bin/python src/python/grpcio_tests/setup.py run_interop --client \
    --args=\"--server_host=grpc-test.sandbox.googleapis.com --server_port=443 --use_tls=true \
    --test_case=empty_unary\""
```

**Comment by jiangtaoli2016** [comment]
In python interop client, we could avoid override if `args.server_host_override` is set to empty string.
https://github.com/grpc/grpc/blob/master/src/python/grpcio_tests/tests/interop/client.py#L103
In this way, at least we could test normal non-override SNI situations.

**Comment by ejona86** [comment]
I've confirmed this bug impacts C++, Python, C#. I think it is safe to assume the other wrapped languages are also impacted.

**Comment by ejona86** [comment]
C++: https://github.com/grpc/grpc/blob/v1.16.0/test/cpp/interop/client.cc#L41
C#: https://github.com/grpc/grpc/blob/v1.16.0/src/csharp/Grpc.IntegrationTesting/InteropClient.cs#L48
Python: https://github.com/grpc/grpc/blob/v1.16.0/src/python/grpcio_tests/tests/interop/client.py#L55
PHP: https://github.com/grpc/grpc/blob/v1.16.0/src/php/tests/interop/interop_client.php#L541
Ruby: https://github.com/grpc/grpc/blob/v1.16.0/src/ruby/pb/test/client.rb#L713

I didn't find any that work doing it correctly. Some I just didn't find the interop client code quickly.

**Comment by jtattermusch** [comment]
CC @jtattermusch

**Comment by jtattermusch** [comment]
Fixed by #17407.
